src/DataStructures/VectorImpl.hpp

```diff
@@ -0,0 +1,410 @@
+// IWYU pragma: no_include <blaze/math/traits/UnaryMapTrait.h>
+// IWYU pragma: no_include <blaze/math/traits/BinaryMapTrait.h>
+#else
+// IWYU pragma: no_include <blaze/math/traits/MapTrait.h>
+#endif  // ((BLAZE_MAJOR_VERSION == 3) && (BLAZE_MINOR_VERSION <= 3))
+// IWYU pragma: no_include <blaze/math/typetraits/TransposeFlag.h>
+
+// IWYU pragma: no_forward_declare blaze::DenseVector
+#if ((BLAZE_MAJOR_VERSION == 3) && (BLAZE_MINOR_VERSION <= 3))
+// IWYU pragma: no_forward_declare blaze::UnaryMapTrait
+// IWYU pragma: no_forward_declare blaze::BinaryMapTrait
+#else
+// IWYU pragma: no_forward_declare blaze::MapTrait
+#endif  // ((BLAZE_MAJOR_VERSION == 3) && (BLAZE_MINOR_VERSION <= 3))
+// IWYU pragma: no_forward_declare blaze::IsVector
+// IWYU pragma: no_forward_declare blaze::TransposeFlag
+
+/*!
+ * \ingroup DataStructuresGroup
+ * \brief Base class template for various DataVector and related types
+ *
+ * \details The `VectorImpl` class is the generic parent class for vectors
+ * representing collections of related function values, such as `DataVector`s
+ * for contiguous data over a computational domain.
+ *
+ * The `VectorImpl` does not itself define any particular mathematical
+ * operations on the contained values. The `VectorImpl` template class and the
+ * macros defined in `VectorImpl.hpp` assist in the construction of various
+ * derived classes supporting a chosen set of mathematical operations.
+ *
+ * In addition, the equivalence operator `==` is inherited from the underlying
+ * `PointerVector` type, and returns true if and only if the size and contents
+ * of the two compared vectors are equivalent.
+ *
+ * Template parameters:
+ * - `T` is the underlying stored type, e.g. `double`, `std::complex<double>`,
+     `float`, etc.
+
+ * - `VectorType` is the type that should be associated with the VectorImpl
+ *    during mathematical computations. In most cases, inherited types should
+ *    have themselves as the second template argument, e.g.
+ *  ```
+ *  class DataVector : VectorImpl<double, DataVector> {
+ *  ```
+ *  The second template parameter communicates arithmetic type restrictions to
+ *  the underlying Blaze framework. For example, if `VectorType` is
+ *  `DataVector`, then the underlying architecture will prevent the addition the
+ *  vector type with `ModalVector`. Since `DataVector`s and `ModalVector`s
+ *  represent data in different spaces, we wish to forbid several operations
+ *  between them. This vector-type-tracking through an expression prevents
+ *  accidental mixing of vector types in math expressions.
+ *
+ * \note
+ * - If created with size 0, then  `data()` will return `nullptr`
+ * - If `SPECTRE_DEBUG` or `SPECTRE_NAN_INIT` macros are defined, then the
+ *   `VectorImpl` is default initialized to `signaling_NaN()`. Otherwise, the
+ *   vector is filled with uninitialized memory for performance.
+ * - Using an object of this type without explicit initialization is undefined
```

Comment:
I find this comment to be confusing rather than helpful, especially as a separate item in a list.   I would remove it.